### PR TITLE
[MLv2] Add functions to support viz settings in FE

### PR DESCRIPTION
**Pull request by bshepherdson**
Viz settings contain legacy MLv1 field refs for the selected columns of
a query. We need to round-trip those settings into MLv2 `:fields` lists
and back to legacy refs. 

The table column picker in the viz settings is migrated to MLv2.

This adds:
- `visible-columns` to get the full set of columns in scope on a query
  - this includes `:selected?` information for the columns
- `legacy-field-ref` to convert an MLv2 field ref to legacy MLv1
- `find-column-index-from-legacy-ref` to match a legacy ref against a
  list of MLv2 columns

Screenshot 2023-08-14 at 18 19 15

### CODE REVIEW

File: frontend/src/metabase-lib/comparison.ts [hunk 1]
```diff
@@ -8,3 +9,17 @@ export function areLegacyQueriesEqual(
 ): boolean {
   return ML.query_EQ_(query1, query2, fieldIds);
 }
+
+export function findColumnIndexForColumnSetting(
+  query: Query,
+  stageIndex: number,
+  columns: ColumnMetadata[],
```
Comment (ranquild) [inline frontend/src/metabase-lib/comparison.ts:16]:
```suggestion
  columns: (ColumnMetadata | DatasetColumn)[],
```

File: frontend/src/metabase-lib/comparison.ts [hunk 2]
```diff
@@ -1,5 +1,6 @@
 import * as ML from "cljs/metabase.lib.js";
-import type { DatasetQuery } from "metabase-types/api";
+import type { DatasetQuery, FieldReference } from "metabase-types/api";
```
Comment (ranquild) [inline frontend/src/metabase-lib/comparison.ts:2]:
```suggestion
import type { DatasetColumn, DatasetQuery, FieldReference } from "metabase-types/api";
```

**Review by ranquild — DISMISSED** [review]

File: src/metabase/lib/js.cljs [hunk 3]
```diff
@@ -418,6 +421,67 @@
...
+;; was difficult to fix entirely within MLv2. Once viz-settings are ported, this function should not be needed, and the
+;; FE logic using it should be ported to MLv2 behind more meaningful names.
+(defn ^:export visible-columns
+  "Return a sequence of column metadatas for columns visible at the given stage of the query. Takes a map of options
+  which defines the set of columns to return. All the options default to *true*, ie. everything is returned by default.
+  ```
+  {
+    include_joined: true,
+    include_expressions true,
+    include_implicitly_joinable: true,
```
Comment (metamben) [inline src/metabase/lib/js.cljs:440]: Shouldn't these be in camelCase?

File: src/metabase/lib/js.cljs [hunk 4]
```diff
@@ -418,6 +421,67 @@
...
+  [a-query stage-number ^js js-options]
+  (let [options     (merge visible-columns-defaults
+                           (when-let [opts (some-> js-options js-keys set)]
+                             (cond-> {}
+                               (opts "includeJoined")
+                               (assoc :include-joined? (.-includeJoined js-options))
+                               (opts "includeExpressions")
+                               (assoc :include-expressions? (.-includeExpressions? js-options))
+                               (opts "includeImplicitlyJoinable")
+                               (assoc :include-implicitly-joinable? (.-includeImplicitlyJoinable js-options)))))
```
Comment (metamben) [inline src/metabase/lib/js.cljs:452]:
```suggestion
  (let [options     (cond-> visible-columns-defaults
                      (gobj/containsKey js-options "includeJoined")
                      (assoc :include-joined? (.-includeJoined js-options))
                      (gobj/containsKey js-options "includeExpressions")
                      (assoc :include-expressions? (.-includeExpressions? js-options))
                      (gobj/containsKey js-options "includeImplicitlyJoinable")
                      (assoc :include-implicitly-joinable? (.-includeImplicitlyJoinable js-options)))
```
Or maybe even just use `(gobj/get js-options "includeJoined" true)` instead of relying on `visible-columns-defaults`.

File: src/metabase/lib/js.cljs [hunk 5]
```diff
@@ -418,6 +421,67 @@
...
+      lib.core/ref
+      convert/->legacy-MBQL
+      (update 2 update-vals #(if (qualified-keyword? %)
+                               (u/qualified-name %)
+                               %))
+      clj->js))
+
+(defn ^:export find-column-index-from-legacy-ref
+  "Given a list of columns (either JS `data.cols` or MLv2 `ColumnMetadata`), search through it for a field matching the
+  given legacy field ref. Returns the index, or -1 if no matching column is found."
```
Comment (metamben) [inline src/metabase/lib/js.cljs:471]: Returning -1 is quite standard (especially in Java and C), but I wonder if returning `nil` would be more comfortable. (The implementation of the function would definitely be a bit more succinct.)

File: src/metabase/lib/js/metadata.cljs [hunk 6]
```diff
@@ -24,7 +25,7 @@
 ;;; where keys are a map of String ID => metadata
 
 (defn- object-get [obj k]
-  (when obj
+  (when (and obj (js-in k obj))
```
Comment (metamben) [inline src/metabase/lib/js/metadata.cljs:28]:
```suggestion
  (when (string? k)
```

File: src/metabase/lib/js/metadata.cljs [hunk 7]
```diff
@@ -82,13 +83,27 @@
...
+(defmulti ^:private rename-key-fn
+  "Returns a function of the keys, either renaming each one or preserving it.
+  If this function returns nil for a given key, the original key is preserved.
+  Use [[excluded-keys]] to drop keys from the input.
+
+  Defaults to nil, which means no renaming is done."
+  identity)
+
+(defmethod rename-key-fn :default [_]
+  nil)
```
Comment (metamben) [inline src/metabase/lib/js/metadata.cljs:95]: Wouldn't it be simpler if it defaulted to `identity`? That seems to be the straightforward way to say "no change".

File: src/metabase/lib/js/metadata.cljs [hunk 8]
```diff
@@ -365,15 +385,15 @@
...
+        :when              (and a-table (= (:db-id a-table) database-id))]
     a-table))
 
 (defn- fields [metadata table-id]
-  (for [[_id field-delay] (some-> metadata :fields deref)
-        :let              [a-field (some-> field-delay deref)]
-        :when             (and a-field (= (:table-id a-field) table-id))]
+  (for [[_id field-delay]  (some-> metadata :fields deref)
+        :let               [a-field (some-> field-delay deref)]
+        :when              (and a-field (= (:table-id a-field) table-id))]
```
Comment (metamben) [inline src/metabase/lib/js/metadata.cljs:396]: Are these changes needed?

**Review by metamben — APPROVED** [review]
I have only some nits.

File: src/metabase/lib/js/metadata.cljs (hunk 6 above)
Comment (bshepherdson) [reply src/metabase/lib/js/metadata.cljs:28]: What? Did the suggestion get mangled?

File: src/metabase/lib/js.cljs (hunk 4 above)
Comment (bshepherdson) [reply src/metabase/lib/js.cljs]: I went with the latter option, that works fairly well.

File: src/metabase/lib/js.cljs (hunk 5 above)
Comment (bshepherdson) [reply src/metabase/lib/js.cljs:461]: It's also conventional in Clojure to return nil rather than empty collections, and to return Clojure collections rather than JS arrays, and we do both those things in this file. This is the JS shim file, and I argue for following JS conventions in here.

File: src/metabase/lib/js/metadata.cljs (hunk 7 above)
Comment (bshepherdson) [reply src/metabase/lib/js/metadata.cljs:94]: Doing that makes the renaming functions harder to write. With this approach they can be maps called like functions, and things that don't appear in the map are preserved.

File: src/metabase/lib/js/metadata.cljs (hunk 8 above)
Comment (bshepherdson) [reply src/metabase/lib/js/metadata.cljs]: No, I'm not sure where those came from. Fixed.

File: src/metabase/lib/js.cljs (hunk 3 above)
Comment (ranquild) [reply src/metabase/lib/js.cljs]: btw @bshepherdson the FE don't use these options anymore, could we drop them entirely? so it would work like all of them are true

File: src/metabase/lib/js/metadata.cljs (hunk 6 above)
Comment (metamben) [reply src/metabase/lib/js/metadata.cljs:27]: I really meant to use `(string? k)` as the test. Wouldn't that be simpler? (The following line Just Works with `obj` being `nil` and when `k` not being a property of `obj`. I think the only difference is, it doesn't blow up when `k` is not a string. Is that undesirable?

File: src/metabase/lib/js.cljs (hunk 5 above)
Comment (metamben) [reply src/metabase/lib/js.cljs:466]: If JS people expect -1 for not found from such a function, then by all means keep it. 👍

File: src/metabase/lib/js/metadata.cljs (hunk 6 above)
Comment (bshepherdson) [reply src/metabase/lib/js/metadata.cljs:27]: Ah, I see. I think back when I made these changes I was trying to debug what turns out to be the `node-repl` being bad. `goog.object` doesn't work properly there, and was throwing in cases where the browser returns `nil` no problem. I think `(string? k)` suffices here, yes.

File: frontend/src/metabase-lib/comparison.ts (hunk 1 above)
Comment (metamben) [reply frontend/src/metabase-lib/comparison.ts]: In fact, it seems I implemented this suggestion...

**Review by snoe — APPROVED** [review]

File: src/metabase/lib/metadata/calculation.cljc [hunk 9]
```diff
@@ -451,17 +451,19 @@
    ReturnedColumnsOptions
    [:map
     ;; these all default to true
-    [:include-joined?              {:optional true} :boolean]
-    [:include-expressions?         {:optional true} :boolean]
-    [:include-implicitly-joinable? {:optional true} :boolean]]])
+    [:include-joined?                              {:optional true} :boolean]
+    [:include-expressions?                         {:optional true} :boolean]
+    [:include-implicitly-joinable?                 {:optional true} :boolean]
+    [:include-implicitly-joinable-for-source-card? {:optional true} :boolean]]])
```
Comment (metamben) [inline src/metabase/lib/metadata/calculation.cljc:457]: I think this flag points to some inconsistency in the product. Apparently, in some places (e.g., viz settings) we want to see implicitly joinable fields for _source cards_, in other places (orderable and breakoutable columns) we don't.

File: test/metabase/lib/card_test.cljc [hunk 10]
```diff
@@ -78,16 +78,3 @@
...
-    (let [venues-query (lib/query
-                        (lib/composed-metadata-provider
-                         (lib.tu/mock-metadata-provider
-                          {:cards [(assoc (:orders lib.tu/mock-cards) :dataset-query lib.tu/venues-query)]})
-                         meta/metadata-provider)
-                        (:orders lib.tu/mock-cards))]
-      (is (= ["ID" "SUBTOTAL" "TOTAL" "TAX" "DISCOUNT" "QUANTITY" "CREATED_AT" "PRODUCT_ID" "USER_ID"]
-             (mapv :name (get-in lib.tu/mock-cards [:orders :result-metadata]))))
-      (is (= ["ID" "SUBTOTAL" "TOTAL" "TAX" "DISCOUNT" "QUANTITY" "CREATED_AT" "PRODUCT_ID" "USER_ID"]
-             (mapv :name (lib/visible-columns venues-query)))))))
```
Comment (metamben) [inline test/metabase/lib/card_test.cljc:93]: @camsaul Removing this test is questionable, but required for this PR to work.

File: test/metabase/lib/stage_test.cljc [hunk 11]
```diff
@@ -178,7 +178,22 @@
...
+                  :display-name             "Name"
+                  :base-type                :type/Text
+                  :lib/source               :source/implicitly-joinable
+                  :lib/desired-column-alias "USERS__via__USER_ID__NAME"}
+                 {:name                     "LAST_LOGIN"
+                  :display-name             "Last Login"
+                  :base-type                :type/DateTime
+                  :lib/source               :source/implicitly-joinable
+                  :lib/desired-column-alias "USERS__via__USER_ID__LAST_LOGIN"}]
                 (lib/visible-columns query)))))))
```
Comment (metamben) [inline test/metabase/lib/stage_test.cljc:197]: @camsaul Reverting this test is questionable, but required for this PR to work. In this case at least I know that issue I fixed in #30950 by removing implicitly joinable fields everywhere is fixed in this PR differently (i.e., only for orderable and breakoutable columns).

**Comment by alxnddr** [comment]
I found a few cases that do not work as expected, please have a look:

1) Nested question
- Create a stuctured question Q1 from the Orders table
- New -> Question -> Saved Questions -> Q1
- In the Notebook mode untick the "Tax" column
- Click "Visualize"
- Try add the "Tax" column in the sidebar

Actual:
The column is added to the table but not removed from the More column section.

2) A saved question joined with a table
- Create a stuctured question Q1 from the Orders table
- New -> Question -> Saved Questions -> Q1
- Join with products 
- Click "Visualize"
- In the sidebar try adding the "ID" column from the "User" table

Actual:
An "ID" table is added to the table but it is a duplicate of the Orders.ID. The column has not been removed from the sidebar.

3) Tree-like self-join
- New -> Question -> Orders
- Select Id and Tax columns only
- Join Orders on Id = UserId
- For the joined table select UserId and Tax columns only
- Click "Visualize"
- In the sidebar try removing "Tax" column

Actual:
After the query reruns, the column "Orders -> Tax" is removed too.

File: src/metabase/lib/convert.cljc [hunk 12]
```diff
@@ -155,20 +174,17 @@
...
                       stage-source-card-id->pMBQL
                       (m/assoc-some :aggregation aggregations :expressions expressions))
             stage (reduce
-                   (fn [stage k]
-                     (if-not (get stage k)
-                       stage
-                       (update stage k ->pMBQL)))
-                   stage
-                   (disj stage-keys :aggregation :expressions))]
+                    (fn [stage k]
```
Comment (camsaul) [inline src/metabase/lib/convert.cljc:182]: old indentation was correct 🙃

**Comment by ranquild** [comment]
> 1) Nested question
> - Create a stuctured question Q1 from the Orders table
> - New -> Question -> Saved Questions -> Q1
> - In the Notebook mode untick the "Tax" column
> - Click "Visualize"
> - Try add the "Tax" column in the sidebar
> 
> Actual:
> The column is added to the table but not removed from the More column section.

The series of the calls is the following:
- When adding a new column (Tax from the above) in the `table.columns` setting, the FE calls `legacyFieldRef`, which returns `["field", "TAX", {"base-type": "type/Float"}]`
- Query is re-run. `/api/dataset` returns `data.cols` with new columns, and for Tax `field_ref` is `['field', 11, null]`
- The FE calls `findColumnIndexesFromLegacyRefs`, passing `data.cols` and field refs from `table.columns`:
```
const datasetIndexes = Lib.findColumnIndexesFromLegacyRefs(
    query,
    STAGE_INDEX,
    datasetColumns,
    fieldRefs,
  );
```

For Tax column the returned index is `-1`, and this makes the FE think that the column is not included in the response.

Expected behavior:
- The correct index should be returned

**Comment by ranquild** [comment]
> 2) A saved question joined with a table
> - Create a stuctured question Q1 from the Orders table
> - New -> Question -> Saved Questions -> Q1
> - Join with products 
> - Click "Visualize"
> - In the sidebar try adding the "ID" column from the "User" table
> 
> Actual:
> An "ID" table is added to the table but it is a duplicate of the Orders.ID. The column has not been removed from the sidebar.

In this case the FE calls `addField` on a column obtained from `visibleColumns` but it has no effect. The query isn't changed.

**Comment by ranquild** [comment]
> 3) Tree-like self-join
> - New -> Question -> Orders
> - Select Id and Tax columns only
> - Join Orders on Id = UserId
> - For the joined table select UserId and Tax columns only
> - Click "Visualize"
> - In the sidebar try removing "Tax" column
> 
> Actual:
> After the query reruns, the column "Orders -> Tax" is removed too.

I managed to reproduce it by removed `Orders -> Tax` and rerunning the query. After that the original `Tax` column coming from the source table disappeared from the settings.

In this case we have the following fields refs in column settings
Screenshot 2023-08-17 at 13 31 10

For some reason `findColumnIndexesFromLegacyRefs` fails to find the correct column for just `Tax` (notice it doesn't have `join-alias` in the image above) but assigns the index for the incorrect field ref with `join-alias`.

`data.cols` here don't have the `join-alias` as expected:

Screenshot 2023-08-17 at 13 37 39

This seems to be another issue with how `findColumnIndexesFromLegacyRefs` finds columns.

File: src/metabase/lib/join/util.cljc [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+    MyJoin__my_field
+
+  You should pass the results thru a unique name function."
+  [query          :- ::lib.schema/query
+   field-metadata :- lib.metadata/ColumnMetadata]
+  (if-let [join-alias (or (current-join-alias field-metadata)
+                          (implicit-join-name query field-metadata))]
+    (joined-field-desired-alias join-alias (:name field-metadata))
+    (:name field-metadata)))
```
Comment (camsaul) [inline src/metabase/lib/join/util.cljc:89]: I don't really agree with moving this here since even a Field without a join alias still has a desired alias

File: test/metabase/lib/js_test.cljs [hunk 14]
```diff
@@ -74,3 +74,7 @@
               {:lib/type :option/join.strategy, :strategy :right-join}
               {:lib/type :option/join.strategy, :strategy :inner-join}]
              (vec strategies))))))
+
+(deftest ^:parallel find-column-index-from-legacy-ref-test
+
+ )
```
Comment (camsaul) [inline test/metabase/lib/js_test.cljs:80]: ????

File: src/metabase/lib/js.cljs [hunk 15]
```diff
@@ -427,6 +431,153 @@
...
+          column->index (into {} (for [index (range (count columns))]
+                                   [(nth columns index) index]))
+          ;; And use that to map each match's ref-index to its column-index.
+          by-index      (into {} (for [[column ref-index] matches]
+                                   [ref-index (column->index column)]))
+          res (->> (range (count legacy-refs))
+                   (map #(by-index % -1))
+                   to-array)]
+      res)))
+
```
Comment (camsaul) [inline src/metabase/lib/js.cljs:580]: I really don't like putting all this logic in `lib.js`, especially without writing any extra tests for it. It's a lot harder to test vs if it's in a `.cljc` file. Plus it's not available if other code needs to build on it in the rest of Metabase lib. 

Something like this should go in a regular `.cljc` namespace with actual tests, with wrappers around it only for doing conversion between JS and Cljs types

File: src/metabase/lib/join/util.cljc (hunk 13 above)
Comment (bshepherdson) [reply src/metabase/lib/join/util.cljc:89]: I'd support an alternative home, it just needed out of `metabase.lib.field` to avoid a circular dep.

`metabase.lib.field.util`?

File: test/metabase/lib/card_test.cljc (hunk 10 above)
Comment (camsaul) [reply test/metabase/lib/card_test.cljc:93]: instead of removing this, can't you call this with the `include-implicitly-joinable-source-card` setting disabled?

File: test/metabase/lib/js_test.cljs (hunk 14 above)
Comment (bshepherdson) [reply test/metabase/lib/js_test.cljs:80]: Leftovers from development that got accidentally committed to the working branch as I was prepping to leave for vacation. There should be some tests written for the custom logic in that CLJS function since it's not a trivial wrapper of another MLv2 function.

File: test/metabase/lib/stage_test.cljc (hunk 11 above)
Comment (camsaul) [reply test/metabase/lib/stage_test.cljc:197]: Can't you just update the call to `visible-columns` to include the new option to exclude implicitly joinable columns for source Cards? Actually it would be even better if you tested both scenarios (with the new option enabled = include the implicitly joinable columns; with the new option disabled = exclude the columns)

File: src/metabase/lib/js.cljs (hunk 15 above)
Comment (bshepherdson) [reply src/metabase/lib/js.cljs:580]: I agree with Cam. Most of this logic looks like it should already have been handled by `lib.equality/find-closest-matches-for-refs`.

File: src/metabase/lib/metadata/calculation.cljc (hunk 9 above)
Comment (metamben) [reply src/metabase/lib/metadata/calculation.cljc:457]: @bshepherdson see @luizarakaki's  [comment](https://github.com/metabase/metabase/issues/30950#issuecomment-1567139855) related to this.

File: test/metabase/lib/card_test.cljc (hunk 10 above)
Comment (metamben) [reply test/metabase/lib/card_test.cljc:93]: That can be done, if we don't really mean what the test was saying.

File: test/metabase/lib/stage_test.cljc (hunk 11 above)
Comment (metamben) [reply test/metabase/lib/stage_test.cljc:197]: That can be done too.

File: src/metabase/lib/js.cljs (hunk 15 above)
Comment (metamben) [reply src/metabase/lib/js.cljs:580]: I does remove part of it, but not the whole. If we want to incorporate something like this in `lib.equality/find-closest-matches-for-refs`, then we'll have to make it optional too and probably disable it at all other call sites. Also note that the composition logic implemented here is a bit different. Here, we make sure that a match early in the series of tests cannot be overridden by a later match. I think `lib.equality/find-closest-matches-for-refs` just assumes there are no duplicate matches. (In fact, I've switched the order of arguments in a `merge` in that function to keep earlier matches, so there might not be a real difference anymore.)

**Comment by bshepherdson** [comment]
I talked this over with @metamben and we have a plan worked out. I'm driving this issue again, now that I'm back.

Next steps are:
- Add some unit tests to cover some missing cases, to increase the safety of the next steps.
- Refactor `find-closest-matches-for-refs` and the extra logic in `metabase.lib.js` to be more unified.
    - Probably also making the "squinting strategy" pluggable, since in some cases we need different precise details of which columns are considered to match, and which aren't.

**Comment by bshepherdson** [comment]
Another odd case:
- Create a saved query for Orders
- Create a new query based on the above.
- Add an implicitly joinable column, eg. `User -> State`
- It works correctly.
- Remove that `User -> State` column again.

Expected:
- `User -> State` disappears from the selected columns
-  And reappears in its original spot in the group of implicitly joinable columns on `Orders.USER_ID`.

Actual:
- It does disappear correctly.
- But it appears under a new `More columns` group separate from the implicit join for `USER_ID`.

**But** if I then re-run the query, it appears correctly. Is this intended? It looks like freshly hidden columns are handled specially, and once the query is re-run and fresh query metadata comes in, the column returns to normal. But I'm not sure if that special handling is deliberate, or naive and missing this case.

**Comment by metamben** [comment]
> Is this intended?

I assumed this was an optimization in the FE so as not to re-run the query for every removed column.

File: src/metabase/lib/js.cljs [hunk 16]
```diff
@@ -465,6 +469,82 @@
...
+    includeJoined: true,
+    includeExpressions true,
+    includeImplicitlyJoinable true,
+  }
+  ```"
+  [a-query stage-number ^js js-options]
+  (let [^js js-options (or js-options #js {})
+        options        {:include-joined?              (gobj/get js-options "includeJoined" true)
+                        :include-expressions?         (gobj/get js-options "includeExpressions" true)
+                        :include-implicitly-joinable? (gobj/get js-options "includeImplicitlyJoinable" true)}
```
Comment (camsaul) [inline src/metabase/lib/js.cljs:489]: would this be easier if we just called `js->clj` on the JS object and then `update-keys` with `u/-:kebab-case-en` and passed that map in directly? Main reason I'm asking is because if we add a new option we'd get it for free that way and not need to hardcode it here as well

File: src/metabase/lib/js.cljs [hunk 17]
```diff
@@ -465,6 +469,82 @@
...
+          ;; First, map each column to its index (in the column list).
+          column->index (into {} (for [index (range (count columns))]
+                                   [(nth columns index) index]))
+          ;; And use that to map each match's ref-index to its column-index.
+          by-index      (into {} (for [[column ref-index] matches]
+                                   [ref-index (column->index column)]))
+          res (->> (range (count legacy-refs))
+                   (map #(by-index % -1))
+                   to-array)]
+      res)))
```
Comment (camsaul) [inline src/metabase/lib/js.cljs:546]: Functions like this should go in the `cljc` files so they're easy to work with and test from a Clojure REPL, stuff in `js.cljs` should just be exports and light type conversion.

File: src/metabase/lib/metadata/calculation.cljc [hunk 18]
```diff
@@ -92,6 +94,9 @@
                          :stage-number stage-number}
                         e)))))))
 
+(defmethod display-name-method :dispatch-type/nil [_query _stage-number _x _stage]
+  "")
```
Comment (camsaul) [inline src/metabase/lib/metadata/calculation.cljc:98]: This seems like kind of an icky hack. Shouldn't the display name of `nil` be `nil`?

File: test/metabase/lib/test_util.cljc [hunk 19]
```diff
@@ -264,6 +280,12 @@
                                        (mapv #(dissoc % :id :table-id)))}))])))
         (meta/tables)))
 
+(defn metadata-provider-with-mock-card [card]
+  (lib/composed-metadata-provider
+    meta/metadata-provider
+    (mock-metadata-provider
+      {:cards [card]})))
```
Comment (camsaul) [inline test/metabase/lib/test_util.cljc:287]: I wrote almost this exact same utility in 

https://github.com/metabase/metabase/pull/33429/files#diff-1da1a58aabf037690d9d1a71c9254ccfdb75eecb7b39f1888f3644ed0d37441aR10-R26

File: src/metabase/lib/js.cljs (hunk 17 above)
Comment (bshepherdson) [reply src/metabase/lib/js.cljs:546]: I'm not sure I agree.

Pushing this logic out to one place isn't so bad. But search for everywhere that `metabase.lib.convert/legacy-ref->pMBQL` is called - that shows dealing with legacy refs escaping from `metabase.lib.js` and finding its way into `lib.field`, `lib.filter`, and one or two more. That feels to me like something that should be dealt with at the edges, not throughout MLv2.

On the other hand, this is mixing JS conversion with some massaging of the return type of `find-closest-matches-for-refs`. I could see moving that inner part into `lib.equality` and adding it to the `discouraged-vars` linter so no one else calls it. That would leave just the JS (and `legacy-ref->pMBQL`) conversions here.

File: src/metabase/lib/js.cljs (hunk 16 above)
Comment (bshepherdson) [reply src/metabase/lib/js.cljs:489]: These options are now never passed as anything but `{}` by the FE, so I've just dropped the argument and all the conversion code.

File: test/metabase/lib/test_util.cljc (hunk 19 above)
Comment (bshepherdson) [reply test/metabase/lib/test_util.cljc:287]: There's a lot of pieces of this branch that will need consolidating. Most of the `lib.js` functionality is partially duplicated from filters work, etc. I'll clean it up in a following PR.

File: src/metabase/lib/metadata/calculation.cljc (hunk 18 above)
Comment (bshepherdson) [reply src/metabase/lib/metadata/calculation.cljc:98]: Removed.

File: src/metabase/lib/js.cljs (hunk 17 above)
Comment (bshepherdson) [reply src/metabase/lib/js.cljs:546]: I've moved the guts of this to `lib.equality` and set that function as a `:discouraged-var`.

**Comment by bshepherdson** [comment]
I've just pushed a new version with 6 squashed, presentable commits in place of the original 57(!) commits. I've still got a side branch with the originals.

It's still broken out by author, and while I'm not sure the tests will pass at each commit the overall flow of the changes makes sense. It's possible there's some auto-merge damage in there somewhere, so we'll see how the CI run goes.

File: src/metabase/lib/equality.cljc [hunk 20]
```diff
@@ -104,106 +109,365 @@
   (mbql.u.match/match-one a-ref
     [:field _opts (_id :guard pos-int?)]))
 
-(defn find-closest-matching-ref
-  "Find the ref that most closely matches `a-ref` from a sequence of `refs`. This is meant to power things
-  like [[metabase.lib.breakout/breakoutable-columns]] which are supposed to include `:breakout-position` for columns
-  that are already present as a breakout; sometimes the column in the breakout does not exactly match what MLv2 would
-  have generated. So try to figure out which column it is referring to.
+(mu/defn resolve-field-id :- lib.metadata/ColumnMetadata
```
Comment (metamben) [inline src/metabase/lib/equality.cljc:112]: I wouldn't be able to guess this is the namespace to look for this function. I would look for it in some metadata namespace.

Comment (bshepherdson) [reply src/metabase/lib/equality.cljc:112]: Yeah, I agree that this location sucks (see the docstring).

It used to be in `lib/field` and only used from there. Now it's used from `lib/field` and `lib/equality`. `lib/field` requires `lib/equality`. `resolve-field-id` requires `lib.card/saved-question-metadata`, and of course everyone, including `lib.card`, requires `lib.metadata.calculation`.

So I can see these options:
1. Move `resolve-field-id`, `lib.card/saved-question-metadata`, and the latter's private helpers to somewhere shared, probably `lib.metadata.calculation`.
2. Move `resolve-field-id` to a new namespace, perhaps `metabase.lib.metadata.resolve` (`resolvers`?)
3. Leave it where it is.

If we're unhappy with where it is in this PR then I think option 2 is the way to go. Thoughts?

File: src/metabase/lib/equality.cljc [hunk 21]
```diff
@@ -104,106 +109,365 @@
...
+           current-haystack        haystack-metadatas
+           [ref-fn & more-ref-fns] ref-fns
+           result                  {}]
+      (cond
+        (or (empty? current-needles)
+            (empty? current-haystack)
+            (nil? ref-fn))            result
+        (nil? ref-fn)                 (recur current-needles current-haystack more-ref-fns result)
+        :else
+        (let [haystack-ref->col  (m/index-by ref-fn current-haystack)
```
Comment (metamben) [inline src/metabase/lib/equality.cljc:343]: Should it be an error if `(< (count haystack-ref->col) (count current-haystack))`?

File: src/metabase/lib/equality.cljc [hunk 22]
```diff
@@ -104,106 +109,365 @@
...
+  - Map `(first ref-fns)` over the `haystack-metadatas`.
+  - Annotate these results with metadata to map them back to the original `haystack-metadatas`.
+  - Call [[find-closest-matches-for-refs]] with `needles` and this list of refs.
+  - Map the results back to use the original `haystack-metadatas` as keys, rather than the refs.
+  - Merge the results right-to-left, so that **earlier matches win**.
+  - Remove any matched `needles` and `haystack-metadatas` from the inputs.
+  - If both inputs are nonempty, continue to the next `ref-fns`, if any."
+  [needles haystack-metadatas opts ref-fns]
+  (let [original-needles (annotate-index needles)]
+    (loop [current-needles         original-needles
```
Comment (metamben) [inline src/metabase/lib/equality.cljc:333]: What about keeping `current-needles` as vectors to make lookup by position faster?

File: src/metabase/lib/equality.cljc [hunk 23]
```diff
@@ -104,106 +109,365 @@
...
+           current-haystack        haystack-metadatas
+           [ref-fn & more-ref-fns] ref-fns
+           result                  {}]
+      (cond
+        (or (empty? current-needles)
+            (empty? current-haystack)
+            (nil? ref-fn))            result
+        (nil? ref-fn)                 (recur current-needles current-haystack more-ref-fns result)
```
Comment (metamben) [inline src/metabase/lib/equality.cljc:341]: The previous line must have matched before this matches. It would be an infinite loop anyway.

File: src/metabase/lib/equality.cljc [hunk 24]
```diff
@@ -104,106 +109,365 @@
...
+        (ffirst at-depth)))))
 
-  This first looks for a matching ref with a strict comparison, then in increasingly less-strict comparisons until it
-  finds something that matches. This is mostly to work around bugs like #31482 where MLv1 generated queries with
-  `:field` refs that did not include join aliases even tho the Fields came from joined Tables... we still know the
-  Fields are the same if they have the same IDs.
+(defn find-closest-matches-for-refs
+  "For each of `needles` (which must be a ref) find the column or ref in `haystack` that it most closely matches. This
+  is meant to power things like [[metabase.lib.breakout/breakoutable-columns]] which are supposed to include
+  `:breakout-position` for columns that are already present as a breakout; sometimes the column in the breakout does not
```
Comment (metamben) [inline src/metabase/lib/equality.cljc:218]: Isn't this a bad example? Is this function recommended when matching against columns?

File: src/metabase/lib/equality.cljc [hunk 25]
```diff
@@ -104,106 +109,365 @@
...
+
+  ([needles haystack-metadatas opts]
+   (when (seq haystack-metadatas)
+     ;; The matching is done in two passes: first as plain refs, second by forcing ID refs.
+     ;; Each of these returns a map of `{^{::index i} haystack-ref needle-index}`. Those haystack values which match
+     ;; are removed between stages.
+     (cascading-matches
+       needles haystack-metadatas opts
+       [;; The first pass attempts to match the refs using the regular [[find-closest-matches-for-refs]].
+        lib.ref/ref
```
Comment (metamben) [inline src/metabase/lib/equality.cljc:384]: I still have the feeling that this can result in a worse match than optimal by ignoring a good match by ID of a column from a source card (generating a ref by name) in favor of a worse match from a table (generating a ref by ID).

File: src/metabase/lib/equality.cljc [hunk 26]
```diff
@@ -104,106 +109,365 @@
...
+         by-name (when (and query
+                            (some field-id-ref? needles))
+                   (closest-matches-in-metadata
+                     (for [[_field _opts field-id] needles]
+                       (when (pos-int? field-id)
+                         (lib.ref/ref (resolve-field-id query stage-number field-id))))
+                     haystack-metadatas
+                     opts))]
+     ;; Prefer the more precise integer ID matches over the name-based ones.
+     (merge by-name basic))))
```
Comment (metamben) [inline src/metabase/lib/equality.cljc:424]: Would you pass in `query` whenever you can? Or are there situations when this could lead to undesired matches?

I have the feeling we have now lots of options and now rules which one to use. I'm afraid we'll be just experimenting which call makes the tests pass...

If you have some good rules, we should put them in the namespace docstring and/or add an explanation to each of these functions about the correct use. (I know there is already some documentation, but I'm not sure it's (precise) enough.)
